asserts/snapasserts/validation_sets.go

```diff
@@ -44,6 +60,45 @@ func (e *ValidationSetsConflictError) Error() string {
 	return buf.String()
 }
 
+// ValidationSetsValidationError describes an error arising
+// from validation of snaps against ValidationSets.
+type ValidationSetsValidationError struct {
+	// snapName -> validationSetKey
+	MissingSnaps       map[string][]string
+	InvalidSnaps       map[string][]string
+	WrongRevisionSnaps map[string][]string
+	// snapName -> revision map for WrongRevisionSnaps
+	Revisions map[string]snap.Revision
+	Sets      map[string]*asserts.ValidationSet
+}
+
+func (e *ValidationSetsValidationError) Error() string {
+	buf := bytes.NewBufferString("validation sets assertions are not met:")
+	printDetails := func(header string, details map[string][]string,
+		printSnap func(snapName string, keys []string) string) {
+		if len(details) == 0 {
+			return
+		}
+		fmt.Fprintf(buf, "\n- %s:", header)
+		for snapName, validationSetKeys := range details {
+			fmt.Fprintf(buf, "\n  - %s", printSnap(snapName, validationSetKeys))
+		}
+	}
+
+	printSnapDetails := func(snapName string, validationSetKeys []string) string {
+		return fmt.Sprintf("%s (%s)", snapName, strings.Join(validationSetKeys, ","))
+	}
+
+	printWrongRevisionDetails := func(snapName string, validationSetKeys []string) string {
+		return fmt.Sprintf("%s (%s), revision %s required", snapName, strings.Join(validationSetKeys, ","), e.Revisions[snapName])
```

Comment:
```
%s (required at revision %s by sets %s,  at revision %s by sets %s)
```
in the normal case the parenthesis would have only the first entry because there is only one revision

Reply:
Done.